File: javascript/solutions/exercise4-error-handling-solution.js
/**
 * Exercise 4: Error Handling in Async Code - SOLUTION
 */

// Simulates an unreliable API call that might fail
function unreliableAPICall(endpoint, shouldFail = false) {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (shouldFail) {
        reject(new Error(`Failed to fetch ${endpoint}`));
      } else {
        resolve({ endpoint, data: 'Success!' });
      }
    }, 300);
  });
}

// Retry an async operation up to maxRetries times
async function retryOperation(operation, maxRetries = 3) {
  let lastError;
  
  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      console.log(`Attempt ${attempt}/${maxRetries}...`);
      const result = await operation();
      console.log('Success!');
      return result;
    } catch (error) {
      lastError = error;
      console.log(`Attempt ${attempt} failed: ${error.message}`);
      if (attempt < maxRetries) {
        console.log('Retrying...');
      }
    }
  }
  
  throw lastError;
}

// Fetch data from multiple endpoints, handling individual failures
async function fetchMultipleEndpoints(endpoints) {
  const results = await Promise.allSettled(
    endpoints.map((endpoint, index) => 
      unreliableAPICall(endpoint, index === 1) // Make the second endpoint fail
    )
  );
  
  const successful = [];
  const failed = [];
  
  results.forEach((result, index) => {
    if (result.status === 'fulfilled') {
      successful.push(result.value);
    } else {
      failed.push({
        endpoint: endpoints[index],
        error: result.reason.message
      });
    }
  });
  
  return { successful, failed };
}

// Chain operations with proper error handling
async function chainOperations(initialValue) {
  try {
    console.log(`Starting with value: ${initialValue}`);
    
    // Step 1: Double the value
    const step1 = await new Promise((resolve) => {
      setTimeout(() => {
        const result = initialValue * 2;
        console.log(`Step 1: ${initialValue} * 2 = ${result}`);
        resolve(result);
      }, 200);
    });
    
    // Step 2: Add 10
    const step2 = await new Promise((resolve) => {
      setTimeout(() => {
        const result = step1 + 10;
        console.log(`Step 2: ${step1} + 10 = ${result}`);
        resolve(result);
      }, 200);
    });
    
    // Step 3: Square the result
    const step3 = await new Promise((resolve) => {
      setTimeout(() => {
        const result = step2 * step2;
        console.log(`Step 3: ${step2} * ${step2} = ${result}`);
        resolve(result);
      }, 200);
    });
    
    return step3;
  } catch (error) {
    throw new Error(`Operation chain failed: ${error.message}`);
  }
}

// Main execution
async function main() {
  console.log('Exercise 4: Error Handling - SOLUTION');
  
  // Test 1: Retry logic
  console.log('\n--- Test 1: Retry Logic ---');
  let attemptCount = 0;
  try {
    const result = await retryOperation(async () => {
      attemptCount++;
      // Fail the first two attempts, succeed on the third
      return unreliableAPICall('/data', attemptCount < 3);
    }, 3);
    console.log('Result:', result);
  } catch (error) {
    console.error('All retries failed:', error.message);
  }
  
  // Test 2: Multiple endpoints with mixed results
  console.log('\n--- Test 2: Multiple Endpoints ---');
  try {
    const endpoints = ['/users', '/posts', '/comments'];
    const results = await fetchMultipleEndpoints(endpoints);
    console.log('Successful requests:', results.successful);
    console.log('Failed requests:', results.failed);
  } catch (error) {
    console.error('Error:', error.message);
  }
  
  // Test 3: Operation chain
  console.log('\n--- Test 3: Operation Chain ---');
  try {
    const result = await chainOperations(10);
    console.log('Final result:', result);
  } catch (error) {
    console.error('Chain failed:', error.message);
  }
}

main();

File: python/solutions/exercise4_error_handling_solution.py
"""
Exercise 4: Error Handling in Async Code - SOLUTION
"""

import asyncio


async def unreliable_api_call(endpoint, should_fail=False):
    """Simulate an unreliable API call."""
    await asyncio.sleep(0.3)
    if should_fail:
        raise Exception(f'Failed to fetch {endpoint}')
    return {'endpoint': endpoint, 'data': 'Success!'}


async def retry_operation(operation, max_retries=3):
    """Retry an async operation up to max_retries times."""
    last_error = None
    
    for attempt in range(1, max_retries + 1):
        try:
            print(f'Attempt {attempt}/{max_retries}...')
            result = await operation()
            print('Success!')
            return result
        except Exception as e:
            last_error = e
            print(f'Attempt {attempt} failed: {e}')
            if attempt < max_retries:
                print('Retrying...')
    
    raise last_error


async def fetch_multiple_endpoints(endpoints):
    """Fetch data from multiple endpoints, handling individual failures."""
    # Create tasks that might fail
    tasks = []
    for i, endpoint in enumerate(endpoints):
        # Make the second endpoint fail
        tasks.append(unreliable_api_call(endpoint, i == 1))
    
    # Gather with return_exceptions to get both results and exceptions
    results = await asyncio.gather(*tasks, return_exceptions=True)
    
    successful = []
    failed = []
    
    for i, result in enumerate(results):
        if isinstance(result, Exception):
            failed.append({
                'endpoint': endpoints[i],
                'error': str(result)
            })
        else:
            successful.append(result)
    
    return {
        'successful': successful,
        'failed': failed
    }


async def chain_operations(initial_value):
    """Chain multiple operations with error handling."""
    try:
        print(f'Starting with value: {initial_value}')
        
        # Step 1: Double the value
        await asyncio.sleep(0.2)
        step1 = initial_value * 2
        print(f'Step 1: {initial_value} * 2 = {step1}')
        
        # Step 2: Add 10
        await asyncio.sleep(0.2)
        step2 = step1 + 10
        print(f'Step 2: {step1} + 10 = {step2}')
        
        # Step 3: Square the result
        await asyncio.sleep(0.2)
        step3 = step2 * step2
        print(f'Step 3: {step2} * {step2} = {step3}')
        
        return step3
    except Exception as e:
        raise Exception(f'Operation chain failed: {e}')


async def main():
    """Main execution function."""
    print('Exercise 4: Error Handling - SOLUTION')
    
    # Test 1: Retry logic
    print('\n--- Test 1: Retry Logic ---')
    attempt_count = [0]
    
    async def failing_operation():
        attempt_count[0] += 1
        return await unreliable_api_call('/data', attempt_count[0] < 3)
    
    try:
        result = await retry_operation(failing_operation, 3)
        print('Result:', result)
    except Exception as e:
        print(f'All retries failed: {e}')
    
    # Test 2: Multiple endpoints with mixed results
    print('\n--- Test 2: Multiple Endpoints ---')
    try:
        endpoints = ['/users', '/posts', '/comments']
        results = await fetch_multiple_endpoints(endpoints)
        print('Successful requests:', results['successful'])
        print('Failed requests:', results['failed'])
    except Exception as e:
        print(f'Error: {e}')
    
    # Test 3: Operation chain
    print('\n--- Test 3: Operation Chain ---')
    try:
        result = await chain_operations(10)
        print(f'Final result: {result}')
    except Exception as e:
        print(f'Chain failed: {e}')


if __name__ == '__main__':
    asyncio.run(main())
